File: serialization/php/json/src/JsonSerializationWriter.php
<?php

namespace Microsoft\Kiota\Serialization\Json;

use DateInterval;
use DateTime;
use DateTimeInterface;
use GuzzleHttp\Psr7\Utils;
use Microsoft\Kiota\Abstractions\Enum;
use Microsoft\Kiota\Abstractions\Serialization\Parsable;
use Microsoft\Kiota\Abstractions\Serialization\SerializationWriter;
use Microsoft\Kiota\Abstractions\Types\Byte;
use Microsoft\Kiota\Abstractions\Types\Date;
use Microsoft\Kiota\Abstractions\Types\Time;
use Psr\Http\Message\StreamInterface;

/**
 * @method onBeforeObjectSerialization(?Parsable $value);
 * @method onStartObjectSerialization(?Parsable $value, SerializationWriter $writer);
 * @method onAfterObjectSerialization(?Parsable $value);
 */
class JsonSerializationWriter implements SerializationWriter
{
    /** @var array<mixed> $writer */
    private array $writer = [];

    /** @var string PROPERTY_SEPARATOR */
    private const PROPERTY_SEPARATOR = ',';

    /** @var callable|null $onStartObjectSerialization */
    private $onStartObjectSerialization;

    /** @var callable|null $onAfterObjectSerialization */
    private $onAfterObjectSerialization;

    /** @var callable|null $onBeforeObjectSerialization */
    private $onBeforeObjectSerialization;

    private function writePropertyName(string $propertyName): void {
        $this->writer []= "\"{$propertyName}\":";
    }

    /**
     * @inheritDoc
     */
    public function writeStringValue(?string $key, ?string $value): void {
        if (!empty($key)) {
            $this->writePropertyName($key);
        }
        $propertyValue = $value !== null ? '"'.addcslashes($value, "\\\r\n\"").'"' : 'null';
        $this->writePropertyValue($key, $propertyValue);
    }

    /**
     * @inheritDoc
     */
    public function writeBooleanValue(?string $key, ?bool $value): void {
        if (!empty($key)) {
            $this->writePropertyName($key);
        }
        $valS = ['false', 'true'];
        $vV= $value === null ? 'null' : $valS[$value];
        $this->writePropertyValue($key, $vV);
    }

    /**
     * @inheritDoc
     */
    public function writeFloatValue(?string $key, ?float $value): void {
        if (!empty($key)) {
            $this->writePropertyName($key);
        }
        $this->writePropertyValue($key, $value);
    }

    /**
     * @inheritDoc
     */
    public function writeIntegerValue(?string $key, ?int $value): void {
        if (!empty($key)) {
            $this->writePropertyName($key);
        }
        $this->writePropertyValue($key, $value);
    }

    /**
     * @inheritDoc
     */
    public function writeDateTimeValue(?string $key, ?DateTime $value): void {
        if (!empty($key)) {
            $this->writePropertyName($key);
        }
        if ($value !== null) {
            $this->writePropertyValue($key, "\"{$value->format(DateTimeInterface::RFC3339)}Z\"");
        } else{
            $this->writePropertyValue($key, 'null');
        }
    }

    /**
     * @param string|null $key
     * @param Date|null $value
     * @return void
     */
    public function writeDateValue(?string $key, ?Date $value): void {
        if (!empty($key)) {
            $this->writePropertyName($key);
        }
        if ($value !== null) {
            $valueString = (string)$value;
            $this->writePropertyValue($key, "\"{$valueString}\"");
        } else {
            $this->writePropertyValue($key, 'null');
        }
    }

    /**
     * @inheritDoc
     */
    public function writeCollectionOfObjectValues(?string $key, ?array $values): void {
        if($key !== null){
            $this->writePropertyName($key);
        }
        if ($values !== null) {
            $this->writer [] = '[';
            foreach ($values as $v) {
                $this->writeObjectValue(null, $v);
                $this->writer [] = self::PROPERTY_SEPARATOR;
            }
            if (count($values) > 0) {
                array_pop($this->writer);
            }
            $this->writer [] = ']';
            if ($key !== null) {
                $this->writer []= self::PROPERTY_SEPARATOR;
            }
        } else {
            $this->writePropertyValue($key, 'null');
        }
    }

    /**
     * @inheritDoc
     */
    public function writeObjectValue(?string $key, $value): void {
        if(!empty($key)) {
            $this->writePropertyName($key);
        }
        if ($value === null) {
            $this->writer []= 'null';
        } else {
            if ($this->onBeforeObjectSerialization !== null) {
                $this->onBeforeObjectSerialization($value);
            }
            $this->writer [] = '{';
            if ($this->onStartObjectSerialization !== null) {
                $this->onStartObjectSerialization($value, $this);
            }
            $value->serialize($this);
            if ($this->writer[count($this->writer) - 1] === ',') {
                array_pop($this->writer);
            }
            if ($this->onAfterObjectSerialization !== null) {
                $this->onAfterObjectSerialization($value);
            }
            $this->writer [] = '}';
        }
        if ($key !== null) {
            $this->writer [] = self::PROPERTY_SEPARATOR;
        }
    }

    /**
     * @inheritDoc
     */
    public function getSerializedContent(): StreamInterface {
        if (count($this->writer) > 0 && $this->writer[count($this->writer) - 1] === ','){
            array_pop($this->writer);
        }
        return Utils::streamFor(implode('', $this->writer));
    }

    /**
     * @inheritDoc
     */
    public function writeEnumSetValue(?string $key, ?array $values): void {
        if (!empty($key)) {
            $this->writePropertyName($key);
        }
        if ($values !== null) {
            $valS = [];
            foreach ($values as $value){
                $valS []= $value->value();
            }
            $this->writeStringValue($key, implode(',', $valS));
        } else {
            $this->writePropertyValue($key, 'null');
        }
    }

    /**
     * @inheritDoc
     */
    public function writeEnumValue(?string $key, ?Enum $value): void {
        if (!empty($key)) {
            $this->writePropertyName($key);
        }
        if ($value !== null) {
            $this->writePropertyValue($key, "\"{$value->value()}\"");
        } else {
            $this->writePropertyValue($key, 'null');
        }
    }

    /**
     * @inheritDoc
     */
    public function writeNullValue(?string $key): void {
        if (!empty($key)) {
            $this->writePropertyName($key);
        }
        $this->writePropertyValue($key, 'null');
    }

    /**
     * @inheritDoc
     */
    public function writeAdditionalData(?array $value): void {
        if($value === null) {
            return;
        }
        foreach ($value as $key => $val) {
            $this->writeAnyValue($key, $val);
        }
    }

    /**
     * @inheritDoc
     */
    public function setOnBeforeObjectSerialization(?callable $value): void {
        $this->onBeforeObjectSerialization = $value;
    }

    /**
     * @inheritDoc
     */
    public function getOnBeforeObjectSerialization(): ?callable {
        return $this->onBeforeObjectSerialization;
    }

    /**
     * @inheritDoc
     */
    public function setOnAfterObjectSerialization(?callable $value): void {
        $this->onAfterObjectSerialization = $value;
    }

    /**
     * @inheritDoc
     */
    public function getOnAfterObjectSerialization(): ?callable {
        return $this->onAfterObjectSerialization;
    }

    /**
     * @inheritDoc
     */
    public function setOnStartObjectSerialization(?callable $value): void {
        $this->onStartObjectSerialization = $value;
    }

    /**
     * @inheritDoc
     */
    public function getOnStartObjectSerialization(): ?callable {
        return $this->onStartObjectSerialization;
    }

    /**
     * @param string|null $key
     * @param mixed $value
     * @throws \JsonException
     */
    public function writeAnyValue(?string $key, $value): void{
        $type = get_debug_type($value);
        switch ($type) {
            case 'float':
                $this->writeFloatValue($key, $value);
                break;
            case 'string':
                $this->writeStringValue($key, $value);
                break;
            case 'int':
                $this->writeIntegerValue($key, $value);
                break;
            case 'bool':
                $this->writeBooleanValue($key, $value);
                break;
            case 'null':
                $this->writeNullValue($key);
                break;
            case Date::class:
                $this->writeDateValue($key, $value);
                break;
            case Time::class:
                $this->writeTimeValue($key, $value);
                break;
            case Byte::class:
                $this->writeByteValue($key, $value);
                break;
            case DateTime::class:
                $this->writeDateTimeValue($key, $value);
                break;
            case DateInterval::class:
                $this->writeDateIntervalValue($key, $value);
                break;
            case 'stdClass':
                $this->writeNonParsableObjectValue($key, (object)$value);
                break;
            case 'array':
                $keys = array_filter(array_keys($value), 'is_string');
                // If there are string keys then that means this is a single
                // object we are dealing with
                // otherwise it is a collection of objects.
                if (!empty($keys)){
                    $this->writeNonParsableObjectValue($key, (object)$value);
                } else if (!empty($value)){
                    if (is_subclass_of($value[0], Parsable::class)) {
                        $this->writeCollectionOfObjectValues($key, $value);
                    } else{
                        $this->writeCollectionOfNonParsableObjectValues($key, $value);
                    }
                }
                break;
            default:
                if (is_a($value, Parsable::class)) {
                    $this->writeObjectValue($key, $value);
                } else if(is_subclass_of($type, Enum::class)){
                    $this->writeEnumValue($key, $value);
                } else if(is_subclass_of($type, DateTimeInterface::class)){
                    $this->writeDateTimeValue($key, $value);
                } else if(is_a($value, StreamInterface::class) || is_subclass_of($type, StreamInterface::class)) {
                    $this->writeStringValue($key, $value->getContents());
                } else {
                   throw new RuntimeException("Could not serialize the object of type {$type}");
                }
                break;
        }
    }

    /**
     * @param string|null $key
     * @param mixed $value
     */
    public function writeNonParsableObjectValue(?string $key, $value): void{
        if(!empty($key)) {
            $this->writePropertyName($key);
        }
        $this->writer []= '{';
        $value = (array)$value;
        foreach ($value as $kKey => $kVal) {
            $this->writeAnyValue($kKey, $kVal);
        }
        if (count($value) > 0) {
            array_pop($this->writer);
        }
        $this->writer []= '}';
        if ($key !== null) {
            $this->writer [] = self::PROPERTY_SEPARATOR;
        }
    }

    /**
     * @param string|null $key
     * @param mixed $value
     * @return void
     */
    private function writePropertyValue(?string $key, $value): void {
        $this->writer []= $value;

        if ($key !== null) {
            $this->writer []= self::PROPERTY_SEPARATOR;
        }
    }

    /**
     * @param string|null $key
     * @param array<mixed> $values
     * @return void
     */
    public function writeCollectionOfNonParsableObjectValues(?string $key, array $values): void {
        if (!empty($key)) {
            $this->writePropertyName($key);
        }
        $this->writer []= '[';
        foreach ($values as $value){
            $this->writeAnyValue(null, $value);
            $this->writer [] = self::PROPERTY_SEPARATOR;
        }
        if (count($values) > 0){
            array_pop($this->writer);
        }
        $this->writer []= ']';
        if ($key !== null) {
            $this->writer [] = self::PROPERTY_SEPARATOR;
        }
    }

    /**
     * @inheritDoc
     */
    public function writeTimeValue(?string $key, ?Time $value): void {
        if (!empty($key)) {
            $this->writePropertyName($key);
        }

        $val = $value !== null ? "\"{$value}\"" : 'null';
        $this->writePropertyValue($key, $val);
    }

    public function writeDateIntervalValue(?string $key, ?DateInterval $value): void {
        if (!empty($key)) {
            $this->writePropertyName($key);
        }

        $res = null;
        if ($value !== null){
            $res = "P{$value->y}Y{$value->y}M{$value->d}DT{$value->h}H{$value->i}M{$value->s}S";
        }
        $val = $res !== null ? "\"{$res}\"" : 'null';
        $this->writePropertyValue($key, $val);
    }

    /**
     * @inheritDoc
     */
    public function writeByteValue(?string $key, ?Byte $value): void {
        if (!empty($key)) {
            $this->writePropertyName($key);
        }

        $val = $value !== null ? (int)(string)($value) : 'null';
        $this->writePropertyValue($key, $val);
    }
}

File: serialization/python/json/serialization/json_serialization_writer.py
import base64
import json
from datetime import date, datetime, time, timedelta
from enum import Enum
from io import BytesIO
from typing import Any, Callable, Dict, List, Optional, TypeVar
from uuid import UUID

from kiota.abstractions.serialization import Parsable, SerializationWriter

T = TypeVar("T")
U = TypeVar("U", bound=Parsable)


class JsonSerializationWriter(SerializationWriter):

    PROPERTY_SEPARATOR: str = ','

    _on_start_object_serialization: Optional[Callable[[Parsable, SerializationWriter], None]] = None

    _on_before_object_serialization: Optional[Callable[[Parsable], None]] = None

    _on_after_object_serialization: Optional[Callable[[Parsable], None]] = None

    def __init__(self):
        self.writer: Dict = {}

    def write_string_value(self, key: Optional[str], value: Optional[str]) -> Optional[str]:
        """Writes the specified string value to the stream with an optional given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            value (Optional[str]): The string value to be written.
        """
        if key and (value or value == ''):  #keeps empty strings as valid values
            self.writer[key] = value
        elif key and not (value or value == ''):
            self.write_null_value(key)
        elif (value or value == '') and not key:
            return value
        return None

    def write_boolean_value(self, key: Optional[str], value: Optional[bool]) -> Optional[bool]:
        """Writes the specified boolean value to the stream with an optional given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            value (Optional[bool]): The boolean value to be written.
        """
        if key and (value or value is False):  # Escape False value from null check
            self.writer[key] = value
        elif key and not (value or value is False):
            self.write_null_value(key)
        elif (value or value is False) and not key:
            return value
        return None

    def write_int_value(self, key: Optional[str], value: Optional[int]) -> Optional[int]:
        """Writes the specified integer value to the stream with an optional given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            value (Optional[int]): The integer value to be written.
        """
        if key and value:
            self.writer[key] = value
        elif key and not value:
            self.write_null_value(key)
        elif value and not key:
            return value
        return None

    def write_float_value(self, key: Optional[str], value: Optional[float]) -> Optional[float]:
        """Writes the specified float value to the stream with an optional given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            value (Optional[float]): The float value to be written.
        """
        if key and value:
            self.writer[key] = value
        elif key and not value:
            self.write_null_value(key)
        elif value and not key:
            return value
        return None

    def write_uuid_value(self, key: Optional[str], value: Optional[UUID]) -> Optional[str]:
        """Writes the specified uuid value to the stream with an optional given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            value (Optional[UUId]): The uuid value to be written.
        """
        if key and value:
            self.writer[key] = str(value)
        elif key and not value:
            self.write_null_value(key)
        elif value and not key:
            return str(value)
        return None

    def write_datetime_offset_value(self, key: Optional[str],
                                    value: Optional[datetime]) -> Optional[str]:
        """Writes the specified datetime offset value to the stream with an optional given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            value (Optional[datetime]): The datetime offset value to be written.
        """
        if key and value:
            self.writer[key] = str(value.isoformat())
        elif key and not value:
            self.write_null_value(key)
        elif value and not key:
            return str(value.isoformat())
        return None

    def write_timedelta_value(self, key: Optional[str],
                              value: Optional[timedelta]) -> Optional[str]:
        """Writes the specified timedelta value to the stream with an optional given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            value (Optional[timedelta]): The timedelta value to be written.
        """
        if key and value:
            self.writer[key] = str(value)
        elif key and not value:
            self.write_null_value(key)
        elif value and not key:
            return str(value)
        return None

    def write_date_value(self, key: Optional[str], value: Optional[date]) -> Optional[str]:
        """Writes the specified date value to the stream with an optional given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            value (Optional[date]): The date value to be written.
        """
        if key and value:
            self.writer[key] = str(value)
        elif key and not value:
            self.write_null_value(key)
        elif value and not key:
            return str(value)
        return None

    def write_time_value(self, key: Optional[str], value: Optional[time]) -> Optional[str]:
        """Writes the specified time value to the stream with an optional given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            value (Optional[time]): The time value to be written.
        """
        if key and value:
            self.writer[key] = str(value)
        elif key and not value:
            self.write_null_value(key)
        elif value and not key:
            return str(value)
        return None

    def write_collection_of_primitive_values(self, key: Optional[str],
                                             values: Optional[List[T]]) -> Optional[List[T]]:
        """Writes the specified collection of primitive values to the stream with an optional
        given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            values (Optional[List[T]]): The collection of primitive values to be written.
        """
        if key and (values or values == []):  # Empty list is meaningful
            self.writer[key] = list(map(lambda x: self.write_any_value(None, x), values))
        elif key and not (values or values == []):
            self.write_null_value(key)
        elif (values or values == []) and not key:
            return list(map(lambda x: self.write_any_value(None, x), values))
        return None

    def write_collection_of_object_values(self, key: Optional[str],
                                          values: Optional[List[U]]) -> Optional[List[Dict]]:
        """Writes the specified collection of model objects to the stream with an optional
        given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            values (Optional[List[U]]): The collection of model objects to be written.
        """
        if key and (values or values == []):
            self.writer[key] = list(
                map(lambda x: self.write_object_value(None, x), values)
            )  # type: ignore
        elif key and not (values or values == []):
            self.write_null_value(key)
        elif (values or values == []) and not key:
            return list(map(lambda x: self.write_object_value(None, x), values))  # type: ignore
        return None

    def write_collection_of_enum_values(self, key: Optional[str],
                                        values: Optional[List[Enum]]) -> Optional[str]:
        """Writes the specified collection of enum values to the stream with an optional given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            values Optional[List[Enum]): The enum values to be written.
        """
        if key and (values or values == []):
            self.writer[key] = ','.join(
                list(map(lambda x: self.write_enum_value(None, x), values))  # type: ignore
            )
        elif key and not (values or values == []):
            self.write_null_value(key)
        elif (values or values == []) and key:
            return ','.join(
                list(map(lambda x: self.write_enum_value(None, x), values))  # type: ignore
            )
        return None

    def write_bytearray_value(self, key: Optional[str], value: bytes) -> Optional[str]:
        """Writes the specified byte array as a base64 string to the stream with an optional
        given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            value (bytes): The byte array to be written.
        """
        if key and value:
            base64_bytes = base64.b64encode(value)
            base64_string = base64_bytes.decode('utf-8')
            self.writer['key'] = base64_string
        elif key and not value:
            self.write_null_value(key)
        elif value and not key:
            base64_bytes = base64.b64encode(value)
            base64_string = base64_bytes.decode('utf-8')
            return base64_string
        return None

    def write_object_value(self, key: Optional[str],
                           value: Optional[U]) -> Optional[Dict[Any, Any]]:
        """Writes the specified model object to the stream with an optional given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            value (Parsable): The model object to be written.
        """
        if key and value:
            if self._on_before_object_serialization:
                self._on_before_object_serialization(value)
            if self._on_start_object_serialization:
                self._on_start_object_serialization(value, self)
            temp_writer = JsonSerializationWriter()
            value.serialize(temp_writer)
            if self._on_after_object_serialization:
                self._on_after_object_serialization(value)
            self.writer[key] = temp_writer.writer
        elif key and not value:
            self.write_null_value(key)
        elif value and not key:
            if self._on_before_object_serialization:
                self._on_before_object_serialization(value)
            if self._on_start_object_serialization:
                self._on_start_object_serialization(value, self)
            temp_writer = JsonSerializationWriter()
            value.serialize(temp_writer)
            if self._on_after_object_serialization:
                self._on_after_object_serialization(value)
            return temp_writer.writer
        return None

    def write_enum_value(self, key: Optional[str], value: Optional[Enum]) -> Optional[str]:
        """Writes the specified enum value to the stream with an optional given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            value (Optional[Enum]): The enum value to be written.
        """
        if key and value:
            self.writer[key] = value.name
        elif key and not value:
            self.write_null_value(key)
        elif value and not key:
            return value.name
        return None

    def write_null_value(self, key: Optional[str]) -> None:
        """Writes a null value for the specified key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
        """
        if key:
            self.writer[key] = None

    def write_additional_data_value(self, value: Dict[str, Any]) -> None:
        """Writes the specified additional data to the stream.
        Args:
            value (Dict[str, Any]): he additional data to be written.
        """
        if value:
            for key, val in value.items():
                self.writer[key] = val

    def get_serialized_content(self) -> BytesIO:
        """Gets the value of the serialized content.
        Returns:
            BytesIO: The value of the serialized content.
        """
        json_string = json.dumps(self.writer)
        self.writer.clear()
        stream = BytesIO(json_string.encode('utf-8'))
        return stream

    def get_on_before_object_serialization(self) -> Optional[Callable[[Parsable], None]]:
        """Gets the callback called before the object gets serialized.
        Returns:
            Optional[Callable[[Parsable], None]]:the callback called before the object
            gets serialized.
        """
        return self._on_before_object_serialization

    def get_on_after_object_serialization(self) -> Optional[Callable[[Parsable], None]]:
        """Gets the callback called after the object gets serialized.
        Returns:
            Optional[Optional[Callable[[Parsable], None]]]: the callback called after the object
            gets serialized.
        """
        return self._on_after_object_serialization

    def get_on_start_object_serialization(
        self
    ) -> Optional[Callable[[Parsable, SerializationWriter], None]]:
        """Gets the callback called right after the serialization process starts.
        Returns:
            Optional[Callable[[Parsable, SerializationWriter], None]]: the callback called
            right after the serialization process starts.
        """
        return self._on_start_object_serialization

    def set_on_before_object_serialization(
        self, value: Optional[Callable[[Parsable], None]]
    ) -> None:
        """Sets the callback called before the objects gets serialized.
        Args:
            value (Optional[Callable[[Parsable], None]]): the callback called before the objects
            gets serialized.
        """
        self._on_before_object_serialization = value

    def set_on_after_object_serialization(
        self, value: Optional[Callable[[Parsable], None]]
    ) -> None:
        """Sets the callback called after the objects gets serialized.
        Args:
            value (Optional[Callable[[Parsable], None]]): the callback called after the objects
            gets serialized.
        """
        self._on_after_object_serialization = value

    def set_on_start_object_serialization(
        self, value: Optional[Callable[[Parsable, SerializationWriter], None]]
    ) -> None:
        """Sets the callback called right after the serialization process starts.
        Args:
            value (Optional[Callable[[Parsable, SerializationWriter], None]]): the callback
            called right after the serialization process starts.
        """
        self._on_start_object_serialization = value

    def write_non_parsable_object_value(self, key: Optional[str], value: T) -> Optional[dict]:
        """Writes the specified value to the stream with an optional given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            value (object): The value to be written.
        """
        if key and value:
            self.writer[key] = value.__dict__
        elif key and not value:
            self.write_null_value(key)
        elif value and not key:
            return value.__dict__
        return None

    def write_any_value(self, key: Optional[str], value: Any) -> Any:
        """Writes the specified value to the stream with an optional given key.
        Args:
            key (Optional[str]): The key to be used for the written value. May be null.
            value Any): The value to be written.
        """
        if key and value:
            value_type = type(value)
            if value_type == bool:
                self.write_boolean_value(key, value)
            elif value_type == str:
                self.write_string_value(key, value)
            elif value_type == int:
                self.write_int_value(key, value)
            elif value_type == float:
                self.write_float_value(key, value)
            elif value_type == UUID:
                self.write_uuid_value(key, value)
            elif value_type == datetime:
                self.write_datetime_offset_value(key, value)
            elif value_type == timedelta:
                self.write_timedelta_value(key, value)
            elif value_type == date:
                self.write_date_value(key, value)
            elif value_type == time:
                self.write_time_value(key, value)
            elif isinstance(value, Enum):
                self.write_enum_value(key, value)
            elif isinstance(value, Parsable):
                self.write_object_value(key, value)
            elif hasattr(value, '__dict__'):
                self.write_non_parsable_object_value(key, value)
            else:
                raise Exception(f"Encountered an unknown type during serialization {value_type}")

        elif key and not value:
            self.write_null_value(key)

        elif value and not key:
            value_type = type(value)
            if value_type == bool:
                return self.write_boolean_value(None, value)
            if value_type == str:
                return self.write_string_value(None, value)
            if value_type == int:
                return self.write_int_value(None, value)
            if value_type == float:
                return self.write_float_value(None, value)
            if value_type == UUID:
                return self.write_uuid_value(None, value)
            if value_type == datetime:
                return self.write_datetime_offset_value(None, value)
            if value_type == timedelta:
                return self.write_timedelta_value(None, value)
            if value_type == date:
                return self.write_date_value(None, value)
            if value_type == time:
                return self.write_time_value(None, value)
            if isinstance(value, Enum):
                return self.write_enum_value(None, value)
            if isinstance(value, Parsable):
                return self.write_object_value(None, value)
            if hasattr(value, '__dict__'):
                return self.write_non_parsable_object_value(None, value)
            raise Exception(f"Encountered an unknown type during serialization {value_type}")
        return None
